compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -536,59 +546,61 @@ object Semantic {
+          Hot
 
         case Cold =>
-          val error = AccessCold(field, source, trace.toVector)
-          Result(Hot, error :: Nil)
+          val error = AccessCold(field, trace.toVector)
+          reporter.report(error)
+          Hot
 
         case ref: Ref =>
           val target = if needResolve then resolve(ref.klass, field) else field
-          val trace1 = trace.add(source)
           if target.is(Flags.Lazy) then
-            given Trace = trace1
             val rhs = target.defTree.asInstanceOf[ValDef].rhs
             eval(rhs, ref, target.owner.asClass, cacheResult = true)
           else
             val obj = ref.objekt
             if obj.hasField(target) then
-              Result(obj.field(target), Nil)
+              obj.field(target)
             else if ref.isInstanceOf[Warm] then
               assert(obj.klass.isSubClass(target.owner))
               if target.is(Flags.ParamAccessor) then
                 // possible for trait parameters
                 // see tests/init/neg/trait2.scala
                 //
                 // return `Hot` here, errors are reported in checking `ThisRef`
-                Result(Hot, Nil)
+                Hot
               else if target.hasSource then
                 val rhs = target.defTree.asInstanceOf[ValOrDefDef].rhs
                 eval(rhs, ref, target.owner.asClass, cacheResult = true)
               else
-                val error = CallUnknown(field, source, trace.toVector)
-                Result(Hot, error :: Nil)
+                val error = CallUnknown(field, trace.toVector)
+                reporter.report(error)
+                Hot
             else
-              val error = AccessNonInit(target, trace.add(source).toVector)
-              Result(Hot, error :: Nil)
+              val error = AccessNonInit(target, trace.toVector)
+              reporter.report(error)
+              Hot
 
         case fun: Fun =>
-          report.error("unexpected tree in selecting a function, fun = " + fun.expr.show, source)
-          Result(Hot, Nil)
+          report.error("unexpected tree in selecting a function, fun = " + fun.expr.show, fun.expr)
+          Hot
 
         case RefSet(refs) =>
-          val resList = refs.map(_.select(field, source))
-          val value2 = resList.map(_.value).join
-          val errors = resList.flatMap(_.errors)
-          Result(value2, errors)
+          refs.map(_.select(field)).join
       }
     }
 
-    def call(meth: Symbol, args: List[ArgInfo], receiver: Type, superType: Type, source: Tree, needResolve: Boolean = true): Contextual[Result] = log("call " + meth.show + ", args = " + args, printer, (_: Result).show) {
-      def checkArgs = args.flatMap(_.promote)
+    def call(meth: Symbol, args: List[ArgInfo], receiver: Type, superType: Type, needResolve: Boolean = true): Contextual[Value] = log("call " + meth.show + ", args = " + args, printer, (_: Value).show) {
+      def checkArgs = args.foreach(_.promote)
```

Comment:
Consider renaming to `promoteArgs`. Also consider parentheses `promoteArgs()` to indicate side effect and Unit return type. But many other methods here have side effects, so maybe not.